# Cleaning_Categorical_Data


### Q1.Import the necessary libraries if you are starting a new notebook. Using the same data as the previous lab: we_fn_use_c_marketing_customer_value_analysis.csv

In [119]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [120]:
customer_df = pd.read_csv('/Users/ayeesha/IH-Labs/lab-cleaning-numerical-data/files_for_lab/we_fn_use_c_marketing_customer_value_analysis.csv')

In [121]:
customer_df.shape

(9134, 24)

In [122]:
cols = []
for column in customer_df.columns:
    cols.append(column.lower())
customer_df.columns = cols

customer_df.columns = customer_df.columns.str.replace(' ', '_')

display(customer_df.head())

customer       state  customer_lifetime_value response  coverage education  \
0  BU79786  Washington              2763.519279       No     Basic  Bachelor   
1  QZ44356     Arizona              6979.535903       No  Extended  Bachelor   
2  AI49188      Nevada             12887.431650       No   Premium  Bachelor   
3  WW63253  California              7645.861827       No     Basic  Bachelor   
4  HB64268  Washington              2813.692575       No     Basic  Bachelor   

  effective_to_date employmentstatus gender  income  ...  \
0           2/24/11         Employed      F   56274  ...   
1           1/31/11       Unemployed      F       0  ...   
2           2/19/11         Employed      F   48767  ...   
3           1/20/11       Unemployed      M       0  ...   
4            2/3/11         Employed      M   43836  ...   

  months_since_policy_inception number_of_open_complaints  number_of_policies  \
0                             5                         0                   1   
1                            42                         0                   8   
2                            38                         0                   2   
3                            65                         0                   7   
4                            44                         0                   1   

      policy_type        policy  renew_offer_type  sales_channel  \
0  Corporate Auto  Corporate L3            Offer1          Agent   
1   Personal Auto   Personal L3            Offer3          Agent   
2   Personal Auto   Personal L3            Offer1          Agent   
3  Corporate Auto  Corporate L2            Offer1    Call Center   
4   Personal Auto   Personal L1            Offer1          Agent   

  total_claim_amount  vehicle_class vehicle_size  
0         384.811147   Two-Door Car      Medsize  
1        1131.464935  Four-Door Car      Medsize  
2         566.472247   Two-Door Car      Medsize  
3         529.881344            SUV      Medsize  
4         138.130879  Four-Door Car      Medsize  

[5 rows x 24 columns]

In [123]:
# convert the 'Date' column to datetime format
customer_df['effective_to_date'] = customer_df['effective_to_date'].astype('datetime64[ns]')
 
# Check the format of 'Date' column
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9134 entries, 0 to 9133
Data columns (total 24 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   customer                       9134 non-null   object        
 1   state                          9134 non-null   object        
 2   customer_lifetime_value        9134 non-null   float64       
 3   response                       9134 non-null   object        
 4   coverage                       9134 non-null   object        
 5   education                      9134 non-null   object        
 6   effective_to_date              9134 non-null   datetime64[ns]
 7   employmentstatus               9134 non-null   object        
 8   gender                         9134 non-null   object        
 9   income                         9134 non-null   int64         
 10  location_code                  9134 non-null   object        
 11  marital_status   

### Q2.Find all of the categorical data. Save it in a categorical_df variable.

In [124]:

categorical_df = customer_df.select_dtypes(object)
categorical_df

customer       state response  coverage education employmentstatus  \
0     BU79786  Washington       No     Basic  Bachelor         Employed   
1     QZ44356     Arizona       No  Extended  Bachelor       Unemployed   
2     AI49188      Nevada       No   Premium  Bachelor         Employed   
3     WW63253  California       No     Basic  Bachelor       Unemployed   
4     HB64268  Washington       No     Basic  Bachelor         Employed   
...       ...         ...      ...       ...       ...              ...   
9129  LA72316  California       No     Basic  Bachelor         Employed   
9130  PK87824  California      Yes  Extended   College         Employed   
9131  TD14365  California       No  Extended  Bachelor       Unemployed   
9132  UP19263  California       No  Extended   College         Employed   
9133  Y167826  California       No  Extended   College       Unemployed   

     gender location_code marital_status     policy_type        policy  \
0         F      Suburban        Married  Corporate Auto  Corporate L3   
1         F      Suburban         Single   Personal Auto   Personal L3   
2         F      Suburban        Married   Personal Auto   Personal L3   
3         M      Suburban        Married  Corporate Auto  Corporate L2   
4         M         Rural         Single   Personal Auto   Personal L1   
...     ...           ...            ...             ...           ...   
9129      M         Urban        Married   Personal Auto   Personal L1   
9130      F      Suburban       Divorced  Corporate Auto  Corporate L3   
9131      M      Suburban         Single  Corporate Auto  Corporate L2   
9132      M      Suburban        Married   Personal Auto   Personal L2   
9133      M      Suburban         Single  Corporate Auto  Corporate L3   

     renew_offer_type sales_channel  vehicle_class vehicle_size  
0              Offer1         Agent   Two-Door Car      Medsize  
1              Offer3         Agent  Four-Door Car      Medsize  
2              Offer1         Agent   Two-Door Car      Medsize  
3              Offer1   Call Center            SUV      Medsize  
4              Offer1         Agent  Four-Door Car      Medsize  
...               ...           ...            ...          ...  
9129           Offer2           Web  Four-Door Car      Medsize  
9130           Offer1        Branch  Four-Door Car      Medsize  
9131           Offer1        Branch  Four-Door Car      Medsize  
9132           Offer3        Branch  Four-Door Car        Large  
9133           Offer4   Call Center   Two-Door Car      Medsize  

[9134 rows x 15 columns]

### Q3.Check for NaN values.


In [125]:
customer_df.isnull().sum()

customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
dtype: int64

### Q4.Check all unique values of columns.

In [126]:
col = []
categorical_df = customer_df.select_dtypes(object)
for col in categorical_df.columns:
    print(col, ': ',categorical_df[col].nunique())
    print(col, ': ',categorical_df[col].unique())

customer :  9134
customer :  ['BU79786' 'QZ44356' 'AI49188' ... 'TD14365' 'UP19263' 'Y167826']
state :  5
state :  ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon']
response :  2
response :  ['No' 'Yes']
coverage :  3
coverage :  ['Basic' 'Extended' 'Premium']
education :  5
education :  ['Bachelor' 'College' 'Master' 'High School or Below' 'Doctor']
employmentstatus :  5
employmentstatus :  ['Employed' 'Unemployed' 'Medical Leave' 'Disabled' 'Retired']
gender :  2
gender :  ['F' 'M']
location_code :  3
location_code :  ['Suburban' 'Rural' 'Urban']
marital_status :  3
marital_status :  ['Married' 'Single' 'Divorced']
policy_type :  3
policy_type :  ['Corporate Auto' 'Personal Auto' 'Special Auto']
policy :  9
policy :  ['Corporate L3' 'Personal L3' 'Corporate L2' 'Personal L1' 'Special L2'
 'Corporate L1' 'Personal L2' 'Special L1' 'Special L3']
renew_offer_type :  4
renew_offer_type :  ['Offer1' 'Offer3' 'Offer2' 'Offer4']
sales_channel :  4
sales_channel :  ['Agent' 'Call Cente

### Q5. Check dtypes. Do they all make sense as categorical data?

In [127]:
categorical_df.dtypes

customer            object
state               object
response            object
coverage            object
education           object
employmentstatus    object
gender              object
location_code       object
marital_status      object
policy_type         object
policy              object
renew_offer_type    object
sales_channel       object
vehicle_class       object
vehicle_size        object
dtype: object

- 'customer' is not a categorical data beacuse it has more number of unique values.
- also 'effective_to_date' is also not a categorical data.It has a values of date so i changed before itself. 

### Q6.Does any column contain alpha and numeric data? Decide how to clean it.

- Customer_id, policy, and renew_offer_type have alpha and numeric characters.
- also has no NULL  values. 


### Q7.Would you choose to do anything else to clean or wrangle the categorical data? Comment your decisions.



No, may be i will drop the customer column ,because it has 9134 unique values. 

### Q8.Compare policy_type and policy. What information is contained in these columns. Can you identify what is important?

- policy_type :  3
- policy_type :  ['Corporate Auto' 'Personal Auto' 'Special Auto']

- policy :  9
- policy :  ['Corporate L3' 'Personal L3' 'Corporate L2' 'Personal L1' 'Special L2'
- 'Corporate L1' 'Personal L2' 'Special L1' 'Special L3']
    
    
    

Probably,I would drop policy, which contains bit more detailed information about the policy which has 3 unique values
as personal, corporate or special. The policy_type column may actually be just enough information on this.
Also the column 'policy_type' already contains part of the information contained in 'policy'. 

### Q9.Check number of unique values in each column, can they be combined in any way to ease encoding? Comment your thoughts and make those changes.

In [129]:
categorical_df.apply(pd.Series.unique)

customer            [BU79786, QZ44356, AI49188, WW63253, HB64268, ...
state               [Washington, Arizona, Nevada, California, Oregon]
response                                                    [No, Yes]
coverage                                   [Basic, Extended, Premium]
education           [Bachelor, College, Master, High School or Bel...
employmentstatus    [Employed, Unemployed, Medical Leave, Disabled...
gender                                                         [F, M]
location_code                                [Suburban, Rural, Urban]
marital_status                            [Married, Single, Divorced]
policy_type             [Corporate Auto, Personal Auto, Special Auto]
policy              [Corporate L3, Personal L3, Corporate L2, Pers...
renew_offer_type                     [Offer1, Offer3, Offer2, Offer4]
sales_channel                       [Agent, Call Center, Web, Branch]
vehicle_class       [Two-Door Car, Four-Door Car, SUV, Luxury SUV,...
vehicle_size        

- 'customer' has 9134 unique values i will drop them.
- 'education'  there are 5 unique values i am going to make them as 4.
- 'policy' i will drop it because has the same information as 'policy type'
- 'vehicle_class ' has 5 unique values i am going to make them as 3.



In [130]:
def combining_unique_val_col(df): 
    df = df.drop(['customer','policy'],axis = 1)
    df['education']= df['education'].replace({'Master':'Graduate', 'Doctor':'Graduate'})
    
    df['vehicle_class']= df['vehicle_class'].replace({'Two-Door Car':'Car', 'Four-Door Car':'Car',
                                                 'SUV':'SUV', 'Luxury SUV':'SUV'})
    
    return df

In [131]:
categorical_df = combining_unique_val_col(categorical_df)
categorical_df

state response  coverage education employmentstatus gender  \
0     Washington       No     Basic  Bachelor         Employed      F   
1        Arizona       No  Extended  Bachelor       Unemployed      F   
2         Nevada       No   Premium  Bachelor         Employed      F   
3     California       No     Basic  Bachelor       Unemployed      M   
4     Washington       No     Basic  Bachelor         Employed      M   
...          ...      ...       ...       ...              ...    ...   
9129  California       No     Basic  Bachelor         Employed      M   
9130  California      Yes  Extended   College         Employed      F   
9131  California       No  Extended  Bachelor       Unemployed      M   
9132  California       No  Extended   College         Employed      M   
9133  California       No  Extended   College       Unemployed      M   

     location_code marital_status     policy_type renew_offer_type  \
0         Suburban        Married  Corporate Auto           Offer1   
1         Suburban         Single   Personal Auto           Offer3   
2         Suburban        Married   Personal Auto           Offer1   
3         Suburban        Married  Corporate Auto           Offer1   
4            Rural         Single   Personal Auto           Offer1   
...            ...            ...             ...              ...   
9129         Urban        Married   Personal Auto           Offer2   
9130      Suburban       Divorced  Corporate Auto           Offer1   
9131      Suburban         Single  Corporate Auto           Offer1   
9132      Suburban        Married   Personal Auto           Offer3   
9133      Suburban         Single  Corporate Auto           Offer4   

     sales_channel vehicle_class vehicle_size  
0            Agent           Car      Medsize  
1            Agent           Car      Medsize  
2            Agent           Car      Medsize  
3      Call Center           SUV      Medsize  
4            Agent           Car      Medsize  
...            ...           ...          ...  
9129           Web           Car      Medsize  
9130        Branch           Car      Medsize  
9131        Branch           Car      Medsize  
9132        Branch           Car        Large  
9133   Call Center           Car      Medsize  

[9134 rows x 13 columns]